c/src/lib/libbsp/arm/shared/stm32fxxxx/can/hal-can.c

```diff
@@ -0,0 +1,664 @@
+/**
+ * @file hal-can.c
+ *
+ * @ingroup can
+ *
+ * @brief CAN driver for the STM32xxxx series processors. Provides a 
+ * CAN driver that registers with cpukit/dev/can as a node in the 
+ * filesystem.
+ *
+ */
+
+/*
+ * Copyright (c) 2015 Vecna Technologies, Inc.
+ *
+ * The license and distribution terms for this file may be
+ * found in the file LICENSE in this distribution or at
+ * http://www.rtems.com/license/LICENSE.
+ */
+
+//================== STMF32 Support Functions =================================
+#include <hal-can.h>
+#include <dev/can/can-internal.h>
+#include <rtems.h>
+#include <hal-utils.h>
+#include <stm32f-processor-specific.h>
+
+#include stm_processor_header(TARGET_STM_PROCESSOR_PREFIX)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, can)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, rcc)
+#include stm_header(TARGET_STM_PROCESSOR_PREFIX, gpio)
+
+#include <math.h>
+#include <rtems/irq-extension.h>
+
+// Internal functions to perform RX and TX
+static int                stm32_can_init              (can_bus * self, long baudRate);
+static int                stm32_can_de_init           (can_bus * self);
+static void               stm32_can_gpio_init         (CAN_Instance canInstance);
+static CAN_Timing_Values  rtems_can_get_timing_values (uint32_t desired_baud);
+static CAN_Status         stm32_can_start_tx          (CAN_HandleTypeDef* hCanHandle, can_msg * msg);
+static CAN_Status         stm32_can_start_rx          (CAN_HandleTypeDef *hCanHandle);
+// Tasks
+static rtems_task         stm32_rx_task               (rtems_task_argument arg);
+static rtems_task         stm32_tx_task               (rtems_task_argument arg);
+// ISRs
+static void               stm32_can_isr               (void * arg);
+static void               stm32_can_rx0_isr           (void * arg);
+static void               stm32_can_rx1_isr           (void * arg);
+
+typedef struct stm32_can_bus stm32_can_bus;
+typedef struct HandleWrapper HandleWrapper;
+
+struct HandleWrapper {
+  CAN_HandleTypeDef handle;
+  stm32_can_bus * bus;
+};
+
+struct stm32_can_bus {
+  can_bus base;
+  HandleWrapper wrapper;
+  CAN_Instance instance;
+  CanTxMsgTypeDef TxMessage;
+  CanRxMsgTypeDef RxMessage;
+};
+
```

Comment:
It's not as simple as moving the struct around. You need a way to go from a CAN_HandleTypeDef back to a stm32_can_bus.

However, I think a function might make this a bit cleaner, though maybe a bit slower too.